dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -230,16 +235,22 @@ class NVDECCache {
     }
 
  private:
-    NVDECCache() {}
+    NVDECCache() {
+      int num_devices = 0;
+      CUDA_CALL(cudaGetDeviceCount(&num_devices));
+      dec_cache.resize(num_devices);
+    }
 
     ~NVDECCache() {
       std::scoped_lock lock(access_lock);
-      for (auto &it : dec_cache) {
-        cuvidDestroyDecoder(it.second.decoder);
+      for (auto &dec_cache_elm : dec_cache) {
+        for (auto &it : dec_cache_elm) {
+          cuvidDestroyDecoder(it.second.decoder);
+        }
       }
     }
-
-    std::unordered_multimap<cudaVideoCodec, DecInstance> dec_cache;
+    using codec_map = std::unordered_multimap<cudaVideoCodec, DecInstance>;
+    std::vector<codec_map> dec_cache;
```

Comment:
I think we should have a mutex per device.

Reply:
Fixed